[#98] Add a flag to redirect HTTP traffic to HTTPS

sestrella opened the pull request:
## what
* Add a flag to redirect HTTP traffic to HTTPS
* The hostname included in the redirection can be customized

## why
* It is a common best practice to redirect HTTP traffic to HTTPS
* This workaround is necessary since Elastic Beanstalk HTTP listener rule can't be changed to a `redirect` action via the [general options](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/command-options-general.html)

## references
* Kudos to @yashshanker who posted the simplified solution
* Closes #98

joe-niland commented on main.tf:
```diff
@@ -1169,3 +1169,30 @@ module "dns_hostname" {
   context = module.this.context
 }
+
+data "aws_lb_listener" "http" {
```
```suggestion
data "aws_lb_listener" "http" {
  count = local.enabled && var.loadbalancer_redirect_http_to_https ? 1 : 0
```

joe-niland commented on main.tf:
```diff
@@ -1169,3 +1169,30 @@ module "dns_hostname" {
+
+resource "aws_lb_listener_rule" "redirect_http_to_https" {
+  count        = var.loadbalancer_redirect_http_to_https ? 1 : 0
+  listener_arn = data.aws_lb_listener.http.arn
```
```suggestion
  listener_arn = join("", data.aws_lb_listener.http.*.arn)
```

joe-niland commented on main.tf:
```diff
@@ -1169,3 +1169,30 @@ module "dns_hostname" {
+}
+
+resource "aws_lb_listener_rule" "redirect_http_to_https" {
+  count        = var.loadbalancer_redirect_http_to_https ? 1 : 0
```
```suggestion
  count        = local.enabled && var.loadbalancer_redirect_http_to_https ? 1 : 0
```

joe-niland commented on main.tf:
```diff
@@ -1169,3 +1169,30 @@ module "dns_hostname" {
 }
+
+data "aws_lb_listener" "http" {
+  load_balancer_arn = var.loadbalancer_is_shared ? var.shared_loadbalancer_arn : aws_elastic_beanstalk_environment.default.0.load_balancers[0]
```
```suggestion
  load_balancer_arn = var.loadbalancer_is_shared ? var.shared_loadbalancer_arn : one(module.elastic_beanstalk_environment.default[*].load_balancers)
```

Pretty sure this works but haven't tested it.

joe-niland reviewed (commented):
Thanks @sestrella 
Looks good. I have a couple of suggestions. 
Let us know when it can be moved out of draft.

sestrella commented:
@joe-niland thank you for the review, I addressed most of the comments, however, I would like to do some testing on a project before flagging this PR as "Ready for Review"

sestrella commented on main.tf:
```diff
@@ -1170,3 +1170,31 @@ module "dns_hostname" {
+
+data "aws_lb_listener" "http" {
+  count             = local.enabled && var.loadbalancer_redirect_http_to_https ? 1 : 0
+  load_balancer_arn = var.loadbalancer_is_shared ? var.shared_loadbalancer_arn : one(aws_elastic_beanstalk_environment.default.0.load_balancers)
```
@joe-niland @joe-niland I tried to avoid hard-coding the `0` index here, however, it didn't work:

```tf
one(aws_elastic_beanstalk_environment.default.*.load_balancers)
```

On the other hand, I noticed the `0` index is hard-coded on the `load_balancers` [output](https://github.com/cloudposse/terraform-aws-elastic-beanstalk-environment/blob/master/outputs.tf#L82), so I'm using the same approach for now.

joe-niland replied:
What was the error you saw with the above?

sestrella replied:
This one:

<img width="865" alt="image" src="https://user-images.githubusercontent.com/2049686/217445506-1c008d9b-fa64-4137-bf7c-3845230cbba8.png">

joe-niland replied:
Sorry this was my mistake - in the example code I was thinking of something else and forgot the indexing was there just because count is used on the eb environment resource. This seems fine now.

sestrella commented:
@joe-niland I did some local testing and it seems to work fine, this is the rule added to the HTTP listener:

<img width="461" alt="image" src="https://user-images.githubusercontent.com/2049686/217442244-985973f2-321b-43af-95c3-3ce47d87a3ca.png">

joe-niland commented:
/rebuild-readme

joe-niland commented:
@sestrella would you mind adding this new var to `https://github.com/cloudposse/terraform-aws-elastic-beanstalk-environment/tree/master/examples/complete` so we can check it with Terratest?

sestrella commented:
@joe-niland done

joe-niland commented:
/test all

joe-niland commented:
/rebuild-readme

joe-niland commented:
@sestrella just a minor thing in the BATS test to look at

joe-niland commented:
Also @sestrella could you please run the following:

```
make
make readme
```

joe-niland commented:
@sestrella or change your PR to allow edits by maintainers: https://github.com/cloudposse/terraform-aws-elastic-beanstalk-environment/actions/runs/4121415576/jobs/7117081418

sestrella commented:
@joe-niland I updated the `README.md` and `docs/terraform.md` files as suggested

joe-niland commented:
/test all

joe-niland commented:
@sestrella can you please update the version constraint in examples/**/versions.tf also?

joe-niland commented:
/test bats

joe-niland commented:
/test readme

joe-niland commented:
/test terratest

joe-niland commented:
@sestrella since you changed the main versions.tf again you'll need to regenerate the readme as well

sestrella commented:
@joe-niland sorry for taking so long updating the README file, I got a weird error running `make readme/build` on Mac with M1:

```
make: *** [/Users/sestrella/code/stackbuilders/terraform-aws-elastic-beanstalk-environment/build-harness/modules/readme/Makefile:32: readme/build] Killed: 9
```

After taking a closer look, I realized that [readme/deps](https://github.com/cloudposse/build-harness/blob/master/modules/readme/Makefile#L31) step downloads the wrong binary for the arm64 architecture:

```
❯ uname -a
Darwin Administrators-MacBook-Pro 22.3.0 Darwin Kernel Version 22.3.0: Thu Jan  5 20:48:54 PST 2023; root:xnu-8792.81.2~2/RELEASE_ARM64_T6000 arm64 arm Darwin
…anstalk-environment on  redirect_http_to_https [?] via 💠 default via ❄️  impure (nix-shell)
➜ file build-harness/vendor/gomplate
build-harness/vendor/gomplate: Mach-O 64-bit x86_64 executable, flags:<NOUNDEFS|DYLDLINK|TWOLEVEL>
…anstalk-environment on  redirect_http_to_https [?] via 💠 default via ❄️  impure (nix-shell)
➜ file build-harness/vendor/terraform-docs
build-harness/vendor/terraform-docs: Mach-O 64-bit x86_64 executable
```


I'm planning to report this issue to [cloudposse/build-harness](https://github.com/cloudposse/build-harness) later

sestrella commented on README.md:
```diff
@@ -248,14 +248,14 @@ Available targets:
 | Name | Version |
 |------|---------|
-| <a name="provider_aws"></a> [aws](#provider\_aws) | >= 3.0 |
+| <a name="provider_aws"></a> [aws](#provider\_aws) | 4.41.0 |
```
@joe-niland maybe you have more insights about where this version constraint comes from? The AWS constrain inside the `versions.tf` file remains the same

joe-niland commented:
/test all

joe-niland commented:
Thanks for your contribution and patience @sestrella.
Released as https://github.com/cloudposse/terraform-aws-elastic-beanstalk-environment/releases/tag/0.49.0

Regarding the aws provider version constraint, was this updated by the make target?

sestrella commented:
@joe-niland thank you for the code review and guidance! About:

> Regarding the aws provider version constraint, was this updated by the make target?

I realized this version was coming from the `.terraform.lock.hcl` generated by `terraform init` so in the end I decided to delete this file and run `make readme/build` again which changed the constraint back to `>= 3.0`

joe-niland commented:
> I realized this version was coming from the `.terraform.lock.hcl` generated by `terraform init` so in the end I decided to delete this file and run `make readme/build` again which changed the constraint back to `>= 3.0`

Thanks for clarifying!
